MAINTAINING.md

```diff
@@ -0,0 +1,147 @@
+# RIOT Review Guidelines
+
+The items presented in the following list should be checked for each PR review,
+in this order. This list is however not exhaustive and defines a rather minimal
+baseline on things that should be checked.
+
+Note: Any of the items in this list might be skipped, if **good** reasons are presented.
+
+Also note: This reviewing order is meant to maximize efficiency and minimize overhead, since failing in step n makes steps n+x obsolete. Maintainers should be aware that starting a review with 4. (Code Style) not only wasts their own time, but also the time of the author
+
+## 1. - Review the fundaments
+
+1.1 Does the reasoning for this PR make sense?
+1.2 Do the concepts used by the PR make sense? 
+1.3 Does the PR break with existing concepts?
+1.4 Is the structure of the PR itself valid?
+1.5 Are the testing criteria clear?
+(1.6 Does this PR respect the rights of previous authors?)
+(1.7 Is the PR a duplicate of another PR?)
+
+
+## 2. - Review the code (line-by-line)
+
+The following list is not exhaustive and addresses the coding issues we have
+regularly seen in the past.
+
+2.1 Check for code duplication
+2.2 Check memory usage
+2.3 Check all code paths
+2.4 Check for API compliance
+2.5 Check for consistent error handling
+2.6 Check scope of variables and functions
+2.7 Check for programming errors
+2.8 Check for code efficiency
+
+
+## 3. - Test the PR
+Run tests to verify the correct behavior (see 1.5), or present **good** reasons
+for skipping some/all tests.
+
+
+## 4. - Review code style and documentation
+
+4.1 Check for sufficient high-level (module-level) documentation
+4.2 Verify function level documentation
+4.3 Are critical/hard to understand parts in the code documented?
+4.4 Check coding style (w.r.t. the coding conventions)
```

Comment:
Maybe also link [best practices](https://github.com/RIOT-OS/RIOT/wiki/Best-Practice-for-RIOT-Programming)?

Reply:
I think this should be moved to part 2. "Review the code line-by-line". I don't think reviewing code that doesn't follow coding conventions is possible. And it's easy to say: your contribution is not compliant with RIOT coding conventions or please run uncrustify.
I would also rename this part "Review documentation" since it's mostly related to this.

Reply:
(+ @jia200x work on automating `uncrustify` should reduce the need to comment on any glaring coding style issues ;-))

Reply:
> But it doesn't make sense to comment on coding style if the overall concept of the PR doesn't make any sense

Part 2 is not on overall concepts but rather on code, that's why coding style should go there.

Reply:
Okay, yes that was my mistake. 2 is about code *correctness*, which groups it rather with (3) than with 4.1-3. But my point stands, that the *subjective** code style should take no precedence over the *objective* code correctness and concepts.

* Yes there is a document that describes the rules we agreed upon for style. Nevertheless, these rules are still based on subjective opinion which a contributor might not share (at least for their day to day work) but needs to adapt to.

Reply:
@aabadie basically I understand the reasoning re efficiency of reviewing. Uncrustifying code as a "first check" is low on maintaining time and people should have done this anyway. It sounds like this is where you're coming from here?
IMO the general philosophy/flow of reviews should be high level -> details. Reason being, that if we do code review, and then the higher level structure of the code or some other design aspects need to be changed, we'll have to do code review again anyway. So, we'll always have to do that code review once (whether it comes first or last), but we'll have to do it twice if design aspects need changing.
I would say that section 2. might be renamed "review design of code" or something like that.

Reply:
> Also, sometimes people submit something where the code guidelines are simply not followed which makes it really difficult/feel rather futile to review. It sounds like this is where you're coming from here?

This is exactly what I have in mind. Why starting to review a code (line-by-line) if it doesn't follow the basic coding conventions of RIOT ? In any case, the contributor will have to, at least, uncrustify it, which might totally change the diff. And previous comments will becomes harder to track via github. Here I make the assumption that several people can review at different moment and also that another pass of review will raise other comments which is generally the case).

So I won't talk of code style in the document, but rather coding conventions and again move it to "review code line-by-line". Maybe this should be done even before looking at the code design. Why ? because this ensures that the contributions "looks like" the existing RIOT's code. And maintainers are used to look at this code, you can better concentrate on the real code design/efficiency problems.

> Maybe giving some general guidance as to how, for example run uncrustify or whatever else.

This would be very useful of course. And should be added to the contributing guidelines.